module-3/lab-supervised-learning-feature-extraction/your-code/main.ipynb

```diff
@@ -12,12 +12,10 @@
   },
```

Comment:
- Your code doesn't check if the M is in the last character only if it's in the string . This time it works because you only have this character otherwise it wouldn't work. You can do something like the following:
```Python
#Your code 
def convert_string_to_numeric(s):
    """
    Convert a string value to numeric. If the last character of the string is `M`, obtain the 
    numeric part of the string, multiply it with 1,000,000, then return the result. Otherwise, 
    convert the string to numeric value and return the result.
    
    Args:
        s: The Reviews score in string format.

    Returns:
        The correct numeric value of the Reviews score.
    """
    if "M" in s:
        num=pd.to_numeric(s.replace("M",""))*1000000
    else:
        num=pd.to_numeric(s)
        
    return num

test_string = '4.0M'

convert_string_to_numeric(test_string) == 4000000
#A possible way of doing it 
def convert_string_to_numeric(s):
    if s[-1] == 'M':
        return float(s[:-1]) * 1000000
    else:
        return float(s)
    return np.NaN
test_string = '4.0M'
convert_string_to_numeric(test_string) == 4000000
```
- In this one "The last step is to apply the function to the Reviews column in the following cell" you don't need a lambda to apply another function: 
```Python 
#Instead of 
google_play['Reviews']=google_play['Reviews'].apply(lambda x: convert_string_to_numeric(x))
# Do something more efficient 
google_play['Reviews'] = google_play.Reviews.apply(convert_string_to_numeric)
```
- Good one 👍 "Step 1: Count the number of positive reviews." 
```Python 
#Your code

def positive_function(x):
    """
    Count how many times the string `Positive` appears in a column (exact string match).
    
    Args:
        x: data column
    
    Returns:
        The number of occurrences of `Positive` in the column data.
    """
    result=np.where(x=="Positive", 1, 0)
    
    return result
```
- In "Bonus", a shorter way :
```Python
#your code 
review_missing_removed['Sentiment Bool']=positive_function(review_missing_removed['Sentiment'])
google_agg=review_missing_removed.groupby('App').agg({'Sentiment Bool':['sum','count']})
#One possible way using rename columns 
google_agg = review_missing_removed.groupby('App').Sentiment.agg([positive_function, 'count']).rename({'positive_function': 'Positive', 'count': 'Total'}, axis='columns')
```
- Bonus: Add a derived column to google_agg that is the ratio of the Positive and the Total columns. Call this column Positive Ratio.
"Make sure to account for the case where the denominator is zero using the np.where() function."
Do consider the zero denominator case:
```Python
#One way 
google_agg['Positive Ratio'] = np.where(google_agg['Total'] !=0, google_agg['Positive'] / google_agg['Total'], 0)
```